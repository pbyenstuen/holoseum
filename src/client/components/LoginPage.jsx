import React, { useState } from "react";
import { useHistory } from "react-router";
import useSubmit from "./useSubmit";
import InputField from "./InputField";

const LoginPage = ({ api, updateUser }) => {
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [inputError, setInputError] = useState("");
    const history = useHistory();

    const { handleSubmit: handleLogin, submitting, error } = useSubmit(
        async () => {
            await api.auth.logIn({ username, password });
        },
        async () => {
            await updateUser();
            history.push("/admin");
        }
    );

    const validateInput = (e) => {
        e.preventDefault();
        if (!username || !password) {
            setInputError("Vennligst fyll inn begge feltene");
            return;
        }
        handleLogin(e);
        setInputError("");
    }

    return (
        <div id="login-page-container">
            <div id="login-card">
                <h2>Admin - Logg Inn</h2>
                <form onSubmit={validateInput}>
                    {submitting && <h4>Vennligst vent...</h4>}
                    {error && <h4>{error.toString()}</h4>}
                    {inputError && <h4>{inputError.toString()}</h4>}
                    <InputField
                        placeholder={"Brukernavn"}
                        value={username}
                        onValueChange={setUsername}
                    />
                    <InputField
                        placeholder={"Passord"}
                        type={"password"}
                        value={password}
                        onValueChange={setPassword}
                    />
                    <button disabled={submitting}>LOGG INN</button>
                </form>
            </div>
        </div>
    );
};

export default LoginPage;