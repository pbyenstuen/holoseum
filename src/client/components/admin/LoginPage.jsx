import React, { useState } from "react";
import { useHistory } from "react-router";
import { Link } from "react-router-dom";
import useSubmit from "../../hooks/useSubmit";
import InputErrorView from "../shared/InputErrorView";
import InputField from "../shared/InputField";

const LoginPage = ({ api, updateUser }) => {
    const [username, setUsername] = useState("");
    const [password, setPassword] = useState("");
    const [inputError, setInputError] = useState("");
    const history = useHistory();

    const { handleSubmit: handleLogin, submitting, error } = useSubmit(
        async () => {
            await api.auth.logIn({ username, password });
        },
        async () => {
            await updateUser();
            history.push("/admin");
        }
    );

    const validateInput = (e) => {
        e.preventDefault();
        if (!username || !password) {
            setInputError("Vennligst fyll inn begge feltene");
            return;
        }
        handleLogin(e);
        setInputError("");
    }

    return (
        <div id="login-page-container">
            <section id="login-card">
                <Link to="/">‚Üê Hovedside</Link>
                <h2>Admin - Logg Inn</h2>
                <form onSubmit={validateInput}>
                    {submitting && <h4>Vennligst vent...</h4>}
                    {error && <InputErrorView error={error}/>}
                    {inputError && <InputErrorView error={inputError}/>}
                    <InputField
                        name={"username"}
                        placeholder={"Brukernavn"}
                        value={username}
                        onValueChange={setUsername}
                    />
                    <InputField
                        name={"password"}
                        placeholder={"Passord"}
                        type={"password"}
                        value={password}
                        onValueChange={setPassword}
                    />
                    <button disabled={submitting}>LOGG INN</button>
                </form>
            </section>
        </div>
    );
};

export default LoginPage;