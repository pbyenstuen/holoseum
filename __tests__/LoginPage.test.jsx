import React from "react";
import ReactDOM from "react-dom";
import { act, Simulate } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import LoginPage from "../src/client/components/LoginPage";

let container;

beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    document.body.removeChild(container);
    container = null;
});

const mountComponent = async (component) => {
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{component}</MemoryRouter>, container);
    });
}

describe("LoginPage", () => {
    it("can submit username and password", async () => {
        const logIn = jest.fn();
        const updateUser = jest.fn();

        const api = {
            auth: {
                logIn
            }
        }

        await mountComponent(<LoginPage api={api} updateUser={updateUser} />);

        Simulate.change(container.querySelector("input[name='username']"), {
            target: { value: "realDonaldTrump" },
        });

        Simulate.change(container.querySelector("input[name='password']"), {
            target: { value: "yourefired" },
        });

        await act(async () => {
            Simulate.submit(container.querySelector("form"));
        });

        expect(logIn).toBeCalledWith({
            username: "realDonaldTrump",
            password: "yourefired",
        });
    });

    it("can show error message", async () => {
        const api = {
            auth: {
                logIn: () => {
                    throw new Error("Feil brukernavn/passord");
                }
            }
        }

        await mountComponent(<LoginPage api={api} />);

        Simulate.change(container.querySelector("input[name='username']"), {
            target: { value: "realDonaldTrump" },
        });

        Simulate.change(container.querySelector("input[name='password']"), {
            target: { value: "yourehired" },
        });

        await act(async () => {
            await Simulate.submit(container.querySelector("form"));
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Error: Feil brukernavn/passord");
    });

    it("can show error message for empty fields", async () => {
        await mountComponent(<LoginPage />);

        await act(async () => {
            await Simulate.submit(container.querySelector("form"));
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h4").textContent).toEqual("Vennligst fyll inn begge feltene");
    });
});