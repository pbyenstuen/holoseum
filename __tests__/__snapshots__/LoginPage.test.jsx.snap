// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage can show error message 1`] = `"<div id=\\"login-page-container\\"><div id=\\"login-card\\"><a href=\\"/\\">← Hovedside</a><h2>Admin - Logg Inn</h2><form><h4>Error: Feil brukernavn/passord</h4><div><label> <input name=\\"username\\" type=\\"text\\" placeholder=\\"Brukernavn\\" value=\\"realDonaldTrump\\"></label></div><div><label> <input name=\\"password\\" type=\\"password\\" placeholder=\\"Passord\\" value=\\"yourehired\\"></label></div><button>LOGG INN</button></form></div></div>"`;

exports[`LoginPage can show error message for empty fields 1`] = `"<div id=\\"login-page-container\\"><div id=\\"login-card\\"><a href=\\"/\\">← Hovedside</a><h2>Admin - Logg Inn</h2><form><h4>Vennligst fyll inn begge feltene</h4><div><label> <input name=\\"username\\" type=\\"text\\" placeholder=\\"Brukernavn\\" value=\\"\\"></label></div><div><label> <input name=\\"password\\" type=\\"password\\" placeholder=\\"Passord\\" value=\\"\\"></label></div><button>LOGG INN</button></form></div></div>"`;
